File: src/theme_explorer.c
#ifndef HAVE_STRLCPY
#include "compat.h"
#endif/*
 * theme_explorer.c - Interactive Theme Explorer Implementation
 *
 * Provides a full-screen TUI panel for browsing and previewing
 * all available color schemes with live preview.
 */

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 600
#endif

#include "theme_explorer.h"
#include "builtin_themes.h"
#include "logger.h"
#include <stdlib.h>
#include <string.h>

#include <ncurses.h>
#include <ctype.h>

// Preview sample text elements
typedef struct {
    const char *label;
    const char *text;
    int color_index;  // Kitty color index (0-15, or special)
} PreviewElement;

// Color indices from Kitty theme format
#define COLOR_FG      -1  // foreground
#define COLOR_RED      1  // color1 - errors
#define COLOR_GREEN    2  // color2 - user/success
#define COLOR_YELLOW   3  // color3 - status/warnings
#define COLOR_BLUE     4  // color4 - headers
#define COLOR_MAGENTA  5  // color5
#define COLOR_CYAN     6  // color6 - assistant
#define COLOR_WHITE    7  // color7
#define COLOR_BRIGHT_BLUE 12  // color12 - tools

// Preview elements to show for each theme
static const PreviewElement preview_elements[] = {
    {"Foreground",  "Main text color for content",      COLOR_FG},
    {"User",        "[User] Hello, how can you help?",  COLOR_GREEN},
    {"Assistant",   "[Assistant] I can help with...",   COLOR_CYAN},
    {"Status",      "[Status] Processing request...",   COLOR_YELLOW},
    {"Error",       "[Error] Something went wrong!",    COLOR_RED},
    {"Tool",        "[Tool] Running bash command...",   COLOR_BRIGHT_BLUE},
    {"Diff Add",    "+ Added line in green",            COLOR_GREEN},
    {"Diff Remove", "- Removed line in red",            COLOR_RED},
};

#define PREVIEW_ELEMENT_COUNT (sizeof(preview_elements) / sizeof(preview_elements[0]))

// Parse a hex color from theme content
static int parse_color_from_content(const char *content, const char *key, int *r, int *g, int *b) {
    char search_key[32];
    snprintf(search_key, sizeof(search_key), "%s ", key);

    const char *line = strstr(content, search_key);
    if (!line) return -1;

    // Find the hex value
    const char *p = line + strlen(search_key);
    while (*p == ' ' || *p == '\t') p++;

    if (*p != '#') return -1;
    p++;

    // Parse 6 hex digits
    if (strlen(p) < 6) return -1;

    char hex[3] = {0};
    hex[0] = p[0]; hex[1] = p[1];
    *r = (int)strtol(hex, NULL, 16);
    hex[0] = p[2]; hex[1] = p[3];
    *g = (int)strtol(hex, NULL, 16);
    hex[0] = p[4]; hex[1] = p[5];
    *b = (int)strtol(hex, NULL, 16);

    return 0;
}

// Convert RGB to ncurses 256-color index
static int rgb_to_256(int r, int g, int b) {
    // Check if grayscale
    int avg = (r + g + b) / 3;
    int r_diff = abs(r - avg);
    int g_diff = abs(g - avg);
    int b_diff = abs(b - avg);

    if (r_diff < 10 && g_diff < 10 && b_diff < 10) {
        int gray = (avg * 23) / 255;
        return 232 + gray;
    }

    // RGB cube
    int ri = (r * 5) / 255;
    int gi = (g * 5) / 255;
    int bi = (b * 5) / 255;
    return 16 + (36 * ri) + (6 * gi) + bi;
}

// Get color index for a preview element from theme content
static int get_theme_color_256(const char *content, int color_idx) {
    char key[16];
    int r = 255, g = 255, b = 255;

    if (color_idx == COLOR_FG) {
        if (parse_color_from_content(content, "foreground", &r, &g, &b) != 0) {
            return 7;  // Default white
        }
    } else {
        snprintf(key, sizeof(key), "color%d", color_idx);
        if (parse_color_from_content(content, key, &r, &g, &b) != 0) {
            // Fallback to basic ANSI colors
            return color_idx;
        }
    }

    return rgb_to_256(r, g, b);
}

// API Functions

int theme_explorer_get_theme_count(void) {
    return (int)built_in_themes_count;
}

const char* theme_explorer_get_theme_name(int index) {
    if (index < 0 || index >= (int)built_in_themes_count) {
        return NULL;
    }
    return built_in_themes[index].name;
}

int theme_explorer_init(ThemeExplorerState *state) {
    if (!state) return -1;

    memset(state, 0, sizeof(ThemeExplorerState));

    // Get terminal dimensions
    int max_y, max_x;
    getmaxyx(stdscr, max_y, max_x);

    // Create a full-screen window
    state->win = newwin(max_y, max_x, 0, 0);
    if (!state->win) {
        LOG_ERROR("[THEME_EXPLORER] Failed to create window");
        return -1;
    }

    state->win_height = max_y;
    state->win_width = max_x;
    state->selected_index = 0;
    state->scroll_offset = 0;
    state->preview_start_col = max_x / 3;  // Left 1/3 for list, right 2/3 for preview

    // Enable keypad for arrow keys
    keypad(state->win, TRUE);

    LOG_INFO("[THEME_EXPLORER] Initialized with %d themes", theme_explorer_get_theme_count());
    return 0;
}

void theme_explorer_cleanup(ThemeExplorerState *state) {
    if (!state) return;

    if (state->win) {
        delwin(state->win);
        state->win = NULL;
    }

    LOG_DEBUG("[THEME_EXPLORER] Cleaned up");
}

// Render the theme list on the left side
static void render_theme_list(ThemeExplorerState *state) {
    int list_width = state->preview_start_col - 1;
    int list_height = state->win_height - 4;  // Reserve space for header/footer
    int theme_count = theme_explorer_get_theme_count();

    // Draw header
    wattron(state->win, A_BOLD);
    mvwprintw(state->win, 1, 2, "Available Themes (%d)", theme_count);
    wattroff(state->win, A_BOLD);

    // Draw separator line
    mvwvline(state->win, 1, state->preview_start_col - 1, ACS_VLINE, state->win_height - 2);

    // Calculate visible range
    int visible_count = list_height;
    if (state->selected_index < state->scroll_offset) {
        state->scroll_offset = state->selected_index;
    } else if (state->selected_index >= state->scroll_offset + visible_count) {
        state->scroll_offset = state->selected_index - visible_count + 1;
    }

    // Draw theme names
    for (int i = 0; i < visible_count && (i + state->scroll_offset) < theme_count; i++) {
        int theme_idx = i + state->scroll_offset;
        const char *name = theme_explorer_get_theme_name(theme_idx);
        if (!name) continue;

        int row = 3 + i;
        int is_selected = (theme_idx == state->selected_index);

        if (is_selected) {
            wattron(state->win, A_REVERSE | A_BOLD);
        }

        // Clear the line first
        wmove(state->win, row, 2);
        for (int j = 0; j < list_width - 3; j++) {
            waddch(state->win, ' ');
        }

        // Print theme name with indicator
        mvwprintw(state->win, row, 2, "%s %s",
                  is_selected ? ">" : " ", name);

        if (is_selected) {
            wattroff(state->win, A_REVERSE | A_BOLD);
        }
    }

    // Draw scroll indicators if needed
    if (state->scroll_offset > 0) {
        mvwaddch(state->win, 2, list_width / 2, ACS_UARROW);
    }
    if (state->scroll_offset + visible_count < theme_count) {
        mvwaddch(state->win, state->win_height - 3, list_width / 2, ACS_DARROW);
    }
}

// Render the preview panel on the right side
static void render_preview(ThemeExplorerState *state) {
    int preview_col = state->preview_start_col + 1;
    int preview_width = state->win_width - preview_col - 2;
    (void)preview_width;

    const char *theme_name = theme_explorer_get_theme_name(state->selected_index);
    if (!theme_name) return;

    // Get theme content
    const char *content = NULL;
    for (size_t i = 0; i < built_in_themes_count; i++) {
        if (strcmp(built_in_themes[i].name, theme_name) == 0) {
            content = built_in_themes[i].content;
            break;
        }
    }
    if (!content) return;

    // Draw preview header
    wattron(state->win, A_BOLD);
    mvwprintw(state->win, 1, preview_col, "Preview: %s", theme_name);
    wattroff(state->win, A_BOLD);

    // Draw preview elements
    int row = 4;
    for (size_t i = 0; i < PREVIEW_ELEMENT_COUNT && row < state->win_height - 3; i++) {
        const PreviewElement *elem = &preview_elements[i];

        // Get the 256-color index for this element
        int color_idx = get_theme_color_256(content, elem->color_index);

        // Use extended color pairs (starting from 100 to avoid conflicts)
        int pair_id = 100 + (int)i;
        init_pair((short)pair_id, (short)color_idx, -1);

        // Draw label
        mvwprintw(state->win, row, preview_col, "%-12s ", elem->label);

        // Draw sample text with theme color
        wattron(state->win, COLOR_PAIR(pair_id));
        wprintw(state->win, "%s", elem->text);
        wattroff(state->win, COLOR_PAIR(pair_id));

        row += 2;
    }

    // Draw color palette preview
    row += 1;
    if (row < state->win_height - 5) {
        mvwprintw(state->win, row, preview_col, "Color Palette:");
        row++;

        // Show colors 0-7 (normal)
        mvwprintw(state->win, row, preview_col, "Normal:  ");
        for (int c = 0; c <= 7; c++) {
            int idx = get_theme_color_256(content, c);
            int pair = 110 + c;
            init_pair((short)pair, (short)idx, -1);
            wattron(state->win, COLOR_PAIR(pair));
            wprintw(state->win, "███");
            wattroff(state->win, COLOR_PAIR(pair));
            waddch(state->win, ' ');
        }
        row++;

        // Show colors 8-15 (bright)
        mvwprintw(state->win, row, preview_col, "Bright:  ");
        for (int c = 8; c <= 15; c++) {
            int idx = get_theme_color_256(content, c);
            int pair = 110 + c;
            init_pair((short)pair, (short)idx, -1);
            wattron(state->win, COLOR_PAIR(pair));
            wprintw(state->win, "███");
            wattroff(state->win, COLOR_PAIR(pair));
            waddch(state->win, ' ');
        }
    }
}

// Draw the full explorer UI
static void render_explorer(ThemeExplorerState *state) {
    werase(state->win);

    // Draw border
    box(state->win, 0, 0);

    // Draw title
    wattron(state->win, A_BOLD);
    mvwprintw(state->win, 0, 2, " Theme Explorer ");
    wattroff(state->win, A_BOLD);

    // Draw help text at bottom
    mvwprintw(state->win, state->win_height - 1, 2,
              " j/↓:Next  k/↑:Prev  Enter:Select  q/ESC:Cancel ");

    // Render components
    render_theme_list(state);
    render_preview(state);

    wrefresh(state->win);
}

ThemeExplorerResult theme_explorer_run(ThemeExplorerState *state) {
    if (!state || !state->win) {
        return THEME_EXPLORER_ERROR;
    }

    int theme_count = theme_explorer_get_theme_count();
    if (theme_count <= 0) {
        LOG_ERROR("[THEME_EXPLORER] No themes available");
        return THEME_EXPLORER_ERROR;
    }

    // Initial render
    render_explorer(state);

    // Event loop
    int running = 1;
    ThemeExplorerResult result = THEME_EXPLORER_CANCELLED;

    while (running) {
        int ch = wgetch(state->win);

        switch (ch) {
            case 'q':
            case 27:  // ESC
                running = 0;
                result = THEME_EXPLORER_CANCELLED;
                break;

            case '\n':
            case '\r':
            case KEY_ENTER:
                // Select current theme
                {
                    const char *name = theme_explorer_get_theme_name(state->selected_index);
                    if (name) {
                        strlcpy(state->selected_theme, name, sizeof(state->selected_theme));
                        result = THEME_EXPLORER_SELECTED;
                        running = 0;
                    }
                }
                break;

            case 'j':
            case KEY_DOWN:
                if (state->selected_index < theme_count - 1) {
                    state->selected_index++;
                    render_explorer(state);
                }
                break;

            case 'k':
            case KEY_UP:
                if (state->selected_index > 0) {
                    state->selected_index--;
                    render_explorer(state);
                }
                break;

            case 'g':
                // gg - go to top
                state->selected_index = 0;
                state->scroll_offset = 0;
                render_explorer(state);
                break;

            case 'G':
                // G - go to bottom
                state->selected_index = theme_count - 1;
                render_explorer(state);
                break;

            case KEY_PPAGE:  // Page Up
                state->selected_index -= (state->win_height - 6);
                if (state->selected_index < 0) state->selected_index = 0;
                render_explorer(state);
                break;

            case KEY_NPAGE:  // Page Down
                state->selected_index += (state->win_height - 6);
                if (state->selected_index >= theme_count) {
                    state->selected_index = theme_count - 1;
                }
                render_explorer(state);
                break;

            case KEY_HOME:
                state->selected_index = 0;
                state->scroll_offset = 0;
                render_explorer(state);
                break;

            case KEY_END:
                state->selected_index = theme_count - 1;
                render_explorer(state);
                break;

            case KEY_RESIZE:
                // Handle terminal resize
                {
                    int max_y, max_x;
                    getmaxyx(stdscr, max_y, max_x);
                    wresize(state->win, max_y, max_x);
                    mvwin(state->win, 0, 0);
                    state->win_height = max_y;
                    state->win_width = max_x;
                    state->preview_start_col = max_x / 3;
                    render_explorer(state);
                }
                break;

            default:
                break;
        }
    }

    return result;
}

const char* theme_explorer_get_selected(ThemeExplorerState *state) {
    if (!state || state->selected_theme[0] == '\0') {
        return NULL;
    }
    return state->selected_theme;
}

File: zig/tui/theme_explorer.zig
//! Theme Explorer - Interactive Theme Selection UI
//!
//! Provides a full-screen TUI panel for browsing and previewing
//! all available color schemes with live preview.

const std = @import("std");
const c = @cImport({
    @cInclude("ncurses.h");
});
const builtin_themes = @import("builtin_themes.zig");
const colorscheme = @import("colorscheme.zig");
const Rgb = colorscheme.Rgb;
const Theme = colorscheme.Theme;

/// Theme explorer result codes
pub const ThemeExplorerResult = enum {
    cancelled, // User cancelled (ESC/q)
    selected, // User selected a theme (Enter)
    error_, // Error occurred
};

/// Preview element for displaying color samples
const PreviewElement = struct {
    label: []const u8,
    text: []const u8,
    color_index: i8, // Kitty color index (0-15, or -1 for foreground)
};

/// Color indices from Kitty theme format
const COLOR_FG: i8 = -1;
const COLOR_RED: i8 = 1; // color1 - errors
const COLOR_GREEN: i8 = 2; // color2 - user/success
const COLOR_YELLOW: i8 = 3; // color3 - status/warnings
const COLOR_BLUE: i8 = 4; // color4 - headers
const COLOR_MAGENTA: i8 = 5; // color5
const COLOR_CYAN: i8 = 6; // color6 - assistant
const COLOR_WHITE: i8 = 7; // color7
const COLOR_BRIGHT_BLUE: i8 = 12; // color12 - tools

/// Preview elements to show for each theme
const PREVIEW_ELEMENTS = [_]PreviewElement{
    .{ .label = "Foreground", .text = "Main text color for content", .color_index = COLOR_FG },
    .{ .label = "User", .text = "[User] Hello, how can you help?", .color_index = COLOR_GREEN },
    .{ .label = "Assistant", .text = "[Assistant] I can help with...", .color_index = COLOR_CYAN },
    .{ .label = "Status", .text = "[Status] Processing request...", .color_index = COLOR_YELLOW },
    .{ .label = "Error", .text = "[Error] Something went wrong!", .color_index = COLOR_RED },
    .{ .label = "Tool", .text = "[Tool] Running bash command...", .color_index = COLOR_BRIGHT_BLUE },
    .{ .label = "Diff Add", .text = "+ Added line in green", .color_index = COLOR_GREEN },
    .{ .label = "Diff Remove", .text = "- Removed line in red", .color_index = COLOR_RED },
};

/// Theme explorer state
pub const ThemeExplorerState = struct {
    win: ?*c.WINDOW = null,
    selected_index: i32 = 0,
    scroll_offset: i32 = 0,
    win_height: i32 = 0,
    win_width: i32 = 0,
    preview_start_col: i32 = 0,
    selected_theme: [64]u8 = undefined,

    pub fn init() ThemeExplorerState {
        var state = ThemeExplorerState{};
        state.selected_theme[0] = 0;
        return state;
    }
};

/// Get the count of available themes
pub fn getThemeCount() i32 {
    return @intCast(builtin_themes.getThemeCount());
}

/// Get the name of a theme by index
pub fn getThemeName(index: i32) ?[]const u8 {
    if (index < 0 or index >= getThemeCount()) return null;
    const theme = builtin_themes.getThemeByIndex(@intCast(index)) orelse return null;
    return theme.name;
}

/// Parse a color from theme content by key
fn parseColorFromContent(content: []const u8, key: []const u8, r: *u8, g: *u8, b: *u8) bool {
    var search_buf: [32]u8 = undefined;
    const search_key = std.fmt.bufPrint(&search_buf, "{s} ", .{key}) catch return false;

    // Find the key in content
    var idx: usize = 0;
    while (idx < content.len) {
        if (std.mem.startsWith(u8, content[idx..], search_key)) {
            const line_start = idx;
            // Find end of line
            var line_end = line_start;
            while (line_end < content.len and content[line_end] != '\n') : (line_end += 1) {}
            const line = content[line_start..line_end];

            // Find hex value after key
            var parts = std.mem.splitScalar(u8, line, ' ');
            _ = parts.next() orelse return false; // Skip key

            while (parts.next()) |part| {
                const trimmed = std.mem.trim(u8, part, &std.ascii.whitespace);
                if (trimmed.len > 0) {
                    if (Rgb.parse(trimmed)) |rgb| {
                        r.* = rgb.r;
                        g.* = rgb.g;
                        b.* = rgb.b;
                        return true;
                    }
                    return false;
                }
            }
            return false;
        }
        idx += 1;
    }

    return false;
}

/// Get 256-color index from theme content for a specific color
fn getThemeColor256(content: []const u8, color_idx: i8) u8 {
    var r: u8 = 255;
    var g: u8 = 255;
    var b: u8 = 255;

    if (color_idx == COLOR_FG) {
        if (!parseColorFromContent(content, "foreground", &r, &g, &b)) {
            return 7; // Default white
        }
    } else {
        var key_buf: [16]u8 = undefined;
        const key = std.fmt.bufPrint(&key_buf, "color{d}", .{color_idx}) catch return @intCast(color_idx);
        if (!parseColorFromContent(content, key, &r, &g, &b)) {
            return @intCast(color_idx); // Fallback to basic ANSI
        }
    }

    const rgb = Rgb.init(r, g, b);
    return rgb.to256ColorIndex();
}

/// Initialize the theme explorer
pub fn init(state: *ThemeExplorerState) bool {
    // Get terminal dimensions
    var max_y: c_int = 0;
    var max_x: c_int = 0;
    _ = c.getmaxyx(c.stdscr, &max_y, &max_x);

    // Create a full-screen window
    state.win = c.newwin(max_y, max_x, 0, 0);
    if (state.win == null) {
        std.log.err("[THEME_EXPLORER] Failed to create window", .{});
        return false;
    }

    state.win_height = max_y;
    state.win_width = max_x;
    state.selected_index = 0;
    state.scroll_offset = 0;
    state.preview_start_col = @divTrunc(max_x, 3); // Left 1/3 for list, right 2/3 for preview

    // Enable keypad for arrow keys
    _ = c.keypad(state.win, true);

    std.log.info("[THEME_EXPLORER] Initialized with {d} themes", .{getThemeCount()});
    return true;
}

/// Clean up theme explorer resources
pub fn cleanup(state: *ThemeExplorerState) void {
    if (state.win) |win| {
        _ = c.delwin(win);
        state.win = null;
    }
    std.log.debug("[THEME_EXPLORER] Cleaned up", .{});
}

/// Render the theme list on the left side
fn renderThemeList(state: *ThemeExplorerState) void {
    const list_width = state.preview_start_col - 1;
    const list_height = state.win_height - 4; // Reserve space for header/footer
    const theme_count = getThemeCount();

    // Draw header
    _ = c.wattron(state.win, c.A_BOLD);
    _ = c.mvwprintw(state.win, 1, 2, "Available Themes (%d)", theme_count);
    _ = c.wattroff(state.win, c.A_BOLD);

    // Draw separator line
    _ = c.mvwvline(state.win, 1, state.preview_start_col - 1, c.ACS_VLINE, state.win_height - 2);

    // Calculate visible range
    const visible_count = list_height;
    if (state.selected_index < state.scroll_offset) {
        state.scroll_offset = state.selected_index;
    } else if (state.selected_index >= state.scroll_offset + visible_count) {
        state.scroll_offset = state.selected_index - visible_count + 1;
    }

    // Draw theme names
    var i: i32 = 0;
    while (i < visible_count and (i + state.scroll_offset) < theme_count) : (i += 1) {
        const theme_idx = i + state.scroll_offset;
        const name = getThemeName(theme_idx) orelse continue;

        const row = 3 + i;
        const is_selected = (theme_idx == state.selected_index);

        if (is_selected) {
            _ = c.wattron(state.win, c.A_REVERSE | c.A_BOLD);
        }

        // Clear the line first
        _ = c.wmove(state.win, row, 2);
        var j: i32 = 0;
        while (j < list_width - 3) : (j += 1) {
            _ = c.waddch(state.win, ' ');
        }

        // Print theme name with indicator
        const indicator: []const u8 = if (is_selected) "> " else "  ";
        _ = c.mvwprintw(state.win, row, 2, "%s%s", indicator.ptr, name.ptr);

        if (is_selected) {
            _ = c.wattroff(state.win, c.A_REVERSE | c.A_BOLD);
        }
    }

    // Draw scroll indicators if needed
    if (state.scroll_offset > 0) {
        _ = c.mvwaddch(state.win, 2, @divTrunc(list_width, 2), c.ACS_UARROW);
    }
    if (state.scroll_offset + visible_count < theme_count) {
        _ = c.mvwaddch(state.win, state.win_height - 3, @divTrunc(list_width, 2), c.ACS_DARROW);
    }
}

/// Render the preview panel on the right side
fn renderPreview(state: *ThemeExplorerState) void {
    const preview_col = state.preview_start_col + 1;
    _ = state.win_width - preview_col - 2; // preview_width (unused but calculated)

    const theme_name = getThemeName(state.selected_index) orelse return;

    // Get theme content
    const content = builtin_themes.getThemeByName(theme_name) orelse return;

    // Draw preview header
    _ = c.wattron(state.win, c.A_BOLD);
    _ = c.mvwprintw(state.win, 1, preview_col, "Preview: %s", theme_name.ptr);
    _ = c.wattroff(state.win, c.A_BOLD);

    // Draw preview elements
    var row: i32 = 4;
    var i: usize = 0;
    while (i < PREVIEW_ELEMENTS.len and row < state.win_height - 3) : (i += 1) {
        const elem = &PREVIEW_ELEMENTS[i];

        // Get the 256-color index for this element
        const color_idx = getThemeColor256(content.content, elem.color_index);

        // Use extended color pairs (starting from 100 to avoid conflicts)
        const pair_id: c_short = @intCast(100 + i);
        _ = c.init_pair(pair_id, @intCast(color_idx), -1);

        // Draw label
        _ = c.mvwprintw(state.win, row, preview_col, "%-12s ", elem.label.ptr);

        // Draw sample text with theme color
        _ = c.wattron(state.win, c.COLOR_PAIR(pair_id));
        _ = c.wprintw(state.win, "%s", elem.text.ptr);
        _ = c.wattroff(state.win, c.COLOR_PAIR(pair_id));

        row += 2;
    }

    // Draw color palette preview
    row += 1;
    if (row < state.win_height - 5) {
        _ = c.mvwprintw(state.win, row, preview_col, "Color Palette:");
        row += 1;

        // Show colors 0-7 (normal)
        _ = c.mvwprintw(state.win, row, preview_col, "Normal:  ");
        var c_idx: i8 = 0;
        while (c_idx <= 7) : (c_idx += 1) {
            const idx = getThemeColor256(content.content, c_idx);
            const pair: c_short = @intCast(110 + c_idx);
            _ = c.init_pair(pair, @intCast(idx), -1);
            _ = c.wattron(state.win, c.COLOR_PAIR(pair));
            _ = c.wprintw(state.win, "███");
            _ = c.wattroff(state.win, c.COLOR_PAIR(pair));
            _ = c.waddch(state.win, ' ');
        }
        row += 1;

        // Show colors 8-15 (bright)
        _ = c.mvwprintw(state.win, row, preview_col, "Bright:  ");
        c_idx = 8;
        while (c_idx <= 15) : (c_idx += 1) {
            const idx = getThemeColor256(content.content, c_idx);
            const pair: c_short = @intCast(110 + c_idx);
            _ = c.init_pair(pair, @intCast(idx), -1);
            _ = c.wattron(state.win, c.COLOR_PAIR(pair));
            _ = c.wprintw(state.win, "███");
            _ = c.wattroff(state.win, c.COLOR_PAIR(pair));
            _ = c.waddch(state.win, ' ');
        }
    }
}

/// Draw the full explorer UI
fn renderExplorer(state: *ThemeExplorerState) void {
    _ = c.werase(state.win);

    // Draw border
    _ = c.box(state.win, 0, 0);

    // Draw title
    _ = c.wattron(state.win, c.A_BOLD);
    _ = c.mvwprintw(state.win, 0, 2, " Theme Explorer ");
    _ = c.wattroff(state.win, c.A_BOLD);

    // Draw help text at bottom
    _ = c.mvwprintw(state.win, state.win_height - 1, 2, " j/↓:Next  k/↑:Prev  Enter:Select  q/ESC:Cancel ");

    // Render components
    renderThemeList(state);
    renderPreview(state);

    _ = c.wrefresh(state.win);
}

/// Run the theme explorer in a modal loop
/// Returns the selected theme name or null if cancelled
pub fn run(state: *ThemeExplorerState) ThemeExplorerResult {
    if (state.win == null) {
        return .error_;
    }

    const theme_count = getThemeCount();
    if (theme_count <= 0) {
        std.log.err("[THEME_EXPLORER] No themes available", .{});
        return .error_;
    }

    // Initial render
    renderExplorer(state);

    // Event loop
    var running = true;
    var result = ThemeExplorerResult.cancelled;

    while (running) {
        const ch = c.wgetch(state.win);

        switch (ch) {
            'q', 27 => { // ESC
                running = false;
                result = .cancelled;
            },
            '\n', '\r', c.KEY_ENTER => { // Enter
                const name = getThemeName(state.selected_index);
                if (name) |n| {
                    const len = @min(n.len, state.selected_theme.len - 1);
                    @memcpy(state.selected_theme[0..len], n[0..len]);
                    state.selected_theme[len] = 0;
                    result = .selected;
                    running = false;
                }
            },
            'j', c.KEY_DOWN => {
                if (state.selected_index < theme_count - 1) {
                    state.selected_index += 1;
                    renderExplorer(state);
                }
            },
            'k', c.KEY_UP => {
                if (state.selected_index > 0) {
                    state.selected_index -= 1;
                    renderExplorer(state);
                }
            },
            'g' => { // gg - go to top (simplified - just 'g')
                state.selected_index = 0;
                state.scroll_offset = 0;
                renderExplorer(state);
            },
            'G' => { // G - go to bottom
                state.selected_index = theme_count - 1;
                renderExplorer(state);
            },
            c.KEY_PPAGE => { // Page Up
                state.selected_index -= (state.win_height - 6);
                if (state.selected_index < 0) state.selected_index = 0;
                renderExplorer(state);
            },
            c.KEY_NPAGE => { // Page Down
                state.selected_index += (state.win_height - 6);
                if (state.selected_index >= theme_count) {
                    state.selected_index = theme_count - 1;
                }
                renderExplorer(state);
            },
            c.KEY_HOME => {
                state.selected_index = 0;
                state.scroll_offset = 0;
                renderExplorer(state);
            },
            c.KEY_END => {
                state.selected_index = theme_count - 1;
                renderExplorer(state);
            },
            c.KEY_RESIZE => {
                // Handle terminal resize
                var max_y: c_int = 0;
                var max_x: c_int = 0;
                _ = c.getmaxyx(c.stdscr, &max_y, &max_x);
                _ = c.wresize(state.win, max_y, max_x);
                _ = c.mvwin(state.win, 0, 0);
                state.win_height = max_y;
                state.win_width = max_x;
                state.preview_start_col = @divTrunc(max_x, 3);
                renderExplorer(state);
            },
            else => {},
        }
    }

    return result;
}

/// Get the name of the selected theme (after run returns selected)
/// Returns null if no theme was selected
pub fn getSelected(state: *ThemeExplorerState) ?[]const u8 {
    if (state.selected_theme[0] == 0) return null;
    return std.mem.sliceTo(&state.selected_theme, 0);
}

/// Convenience function to run the theme explorer and return selected theme name
/// Returns null if cancelled or error
pub fn showThemeExplorer() ?[]const u8 {
    var state = ThemeExplorerState.init();

    if (!init(&state)) {
        return null;
    }
    defer cleanup(&state);

    const result = run(&state);
    if (result == .selected) {
        return getSelected(&state);
    }

    return null;
}

// Tests
const testing = std.testing;

test "getThemeCount" {
    try testing.expect(getThemeCount() > 0);
}

test "getThemeName" {
    const name = getThemeName(0);
    try testing.expect(name != null);
    try testing.expect(name.?.len > 0);

    try testing.expect(getThemeName(-1) == null);
    try testing.expect(getThemeName(1000) == null);
}

test "parseColorFromContent" {
    const content = "foreground #FF5500\ncolor1 #00AA00\n";
    var r: u8 = 0;
    var g: u8 = 0;
    var b: u8 = 0;

    try testing.expect(parseColorFromContent(content, "foreground", &r, &g, &b));
    try testing.expectEqual(@as(u8, 255), r);
    try testing.expectEqual(@as(u8, 85), g);
    try testing.expectEqual(@as(u8, 0), b);
}

test "getThemeColor256" {
    const content = "color2 #00FF00\nforeground #FFFFFF\n";

    const green_idx = getThemeColor256(content, 2);
    try testing.expect(green_idx >= 16 and green_idx <= 255);

    const white_idx = getThemeColor256(content, -1);
    try testing.expect(white_idx >= 232 or white_idx < 16);
}
